lib/remote_node.go

```diff
@@ -0,0 +1,744 @@
+			return fmt.Errorf("RemoteNode.HandleVersionMessage: Requesting disconnect for id: (%v). Version too low. "+
+				"Outbound RemoteNodes must use at least ProtocolVersion2, instead received version: %v", rn.id, verMsg.Version)
+		}
+
+		negotiatedVersion = NewProtocolVersionType(verMsg.Version)
+	}
+
+	vMeta.negotiatedProtocolVersion = negotiatedVersion
+
+	// Record the services the peer is advertising.
+	vMeta.serviceFlag = verMsg.Services
+	// If the RemoteNode was connected with an expectation of being a validator, make sure that its advertised ServiceFlag
+	// indicates that it is a validator.
+	if !rn.hasValidatorServiceFlag() && rn.validatorPublicKey != nil {
+		return fmt.Errorf("RemoteNode.HandleVersionMessage: Requesting disconnect for id: (%v). "+
+			"Expected validator, but received invalid ServiceFlag: %v", rn.id, verMsg.Services)
+	}
+	// If the RemoteNode is on ProtocolVersion1, then it must not have the validator service flag set.
+	if rn.hasValidatorServiceFlag() && vMeta.advertisedProtocolVersion.Before(ProtocolVersion2) {
+		return fmt.Errorf("RemoteNode.HandleVersionMessage: Requesting disconnect for id: (%v). "+
+			"RemoteNode has SFValidator service flag, but doesn't have ProtocolVersion2 or later", rn.id)
+	}
+
+	// Record the tstamp sent by the peer and calculate the time offset.
+	timeConnected := time.Unix(verMsg.TstampSecs, 0)
+	vMeta.timeConnected = &timeConnected
+	currentTime := time.Now().Unix()
+	if currentTime > verMsg.TstampSecs {
+		vMeta.timeOffsetSecs = uint64(currentTime - verMsg.TstampSecs)
+	} else {
+		vMeta.timeOffsetSecs = uint64(verMsg.TstampSecs - currentTime)
+	}
+
+	// Save the received version nonce so we can include it in our verack message.
+	vMeta.versionNonceReceived = verMsg.Nonce
+
+	// Set the peer info-related fields.
+	vMeta.userAgent = verMsg.UserAgent
+	vMeta.latestBlockHeight = verMsg.LatestBlockHeight
+	vMeta.minTxFeeRateNanosPerKB = verMsg.MinFeeRateNanosPerKB
+
+	// Respond to the version message if this is an inbound peer.
+	if rn.IsInbound() {
+		if err := rn.sendVersionMessage(responseNonce); err != nil {
+			return errors.Wrapf(err, "RemoteNode.HandleVersionMessage: Problem sending version message to peer (id= %d)", rn.id)
+		}
+	}
+
+	// After sending and receiving a compatible version, send the verack message. Notice that we don't wait for the
+	// peer's verack message even if it is an inbound peer. Instead, we just send the verack message right away.
+
+	// Set the latest time by which we should receive a verack message from the peer.
+	verackTimeExpected := time.Now().Add(rn.params.VerackNegotiationTimeout)
+	rn.verackTimeExpected = &verackTimeExpected
+	if err := rn.sendVerack(); err != nil {
+		return errors.Wrapf(err, "RemoteNode.HandleVersionMessage: Problem sending verack message to peer (id= %d)", rn.id)
+	}
+
+	// Update the timeSource now that we've gotten a version message from the peer.
+	rn.cmgr.AddTimeSample(rn.peer.Address(), timeConnected)
```

Comment:
It's fine to leave this here for legacy reasons but we don't want to rely on this timeSource anymore. Leave a comment explaining this and make sure we're not using the timeSource anywhere.